[ASTGen] Generate pattern bindings, accessors, and subscripts

hamishknight opened the pull request:
Complete the generation for pattern bindings, and implement generation for accessors and subscripts.

hamishknight commented:
@swift-ci please test

hamishknight commented on lib/ASTGen/Sources/ASTGen/ParameterClause.swift:
```diff
@@ -61,40 +61,63 @@ extension EnumCaseParameterSyntax: ValueParameterSyntax {
+  /// Generate a ParamDecl. If `argNameByDefault` is true, then the parameter's
+  /// argument label is inferred from the first name if no second name is present.
+  private func makeParamDecl(_ node: some ValueParameterSyntax, argNameByDefault: Bool) -> BridgedParamDecl {
+    precondition(
+      !(node.optionalFirstName == nil && node.secondName != nil),
+      "If only one name is present, it should be the first"
+    )
```
Is this something we can assert, or should we be defensive here? I believe this should only apply to EnumCaseParameterSyntax

rintaro commented on lib/ASTGen/Sources/ASTGen/ParameterClause.swift:
```diff
@@ -121,11 +147,32 @@ extension ASTGenVisitor {
-  func generate(functionParameterList node: FunctionParameterListSyntax) -> BridgedArrayRef {
-    node.lazy.map(self.generate).bridgedArray(in: self)
+  func generate(functionParameterList node: FunctionParameterListSyntax, forSubscript: Bool) -> BridgedArrayRef {
+    node.lazy.map { self.generate(functionParameter: $0, forSubscript: forSubscript) }.bridgedArray(in: self)
```
Just a style preference, for method chaining like this, I'd prefer enclosing the closure in parentheses `map({ ... }).bridgedArray`

rintaro commented on test/ASTGen/verify-parse.swift:
```diff
@@ -1,7 +1,12 @@
+
+// Filter out any addresses in the dump, since they can differ.
+// RUN: sed -E 's#0x[0-9a-fA-F]+##g' %t/cpp-parser.ast.raw > %t/cpp-parser.ast
+// RUN: sed -E 's#0x[0-9a-fA-F]+##g' %t/astgen.ast.raw > %t/astgen.ast
```
Since this test is getting bigger, maybe it makes sense to separate it.
Like `test/ASTGen/decls.swift`, `test/ASTGen/stmts.swift`, ...
That also mitigates the chance of conflicts.

hamishknight commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -288,23 +288,215 @@ extension ASTGenVisitor {
+    }
+    if let accessors = binding.accessorBlock {
+      // FIXME: We ought to have a better way of converting from the Nullable
+      // wrapper back to Optional.
```
Yeah

hamishknight commented on lib/AST/ASTBridging.cpp:
```diff
@@ -324,55 +333,79 @@ void BridgedDiagnostic_finish(BridgedDiagnostic cDiag) {
 
-  VarDecl::Introducer introducer =
-      isLet ? VarDecl::Introducer::Let : VarDecl::Introducer::Var;
+  SmallVector<PatternBindingEntry, 4> entries;
+  for (auto &entry : cBindingEntries.unbridged<BridgedPatternBindingEntry>()) {
```
Hmm, what's wrong with the temporary SmallVector? Wouldn't using `AllocateTransform` cause us to allocate permanent memory that then gets thrown away?

rintaro replied:
Oh, I see `PatternBindingDecl::create()` copies it. Never mind.

hamishknight commented on lib/AST/Decl.cpp:
```diff
@@ -1937,6 +1928,10 @@ PatternBindingDecl::create(ASTContext &Ctx, SourceLoc StaticLoc,
     auto &newEntry = entries[elt];
```
Oof yeah, I'll clean this up while I'm here

hamishknight commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -288,23 +288,215 @@ extension ASTGenVisitor {
+      // FIXME: We ought to have a better way of converting from the Nullable
+      // wrapper back to Optional.
+      if let primaryVar = pattern.singleVar.raw.map(BridgedVarDecl.init) {
+        generate(accessorBlock: accessors, for: primaryVar.asAbstractStorageDecl)
```
Yeah could do, I wasn't convinced it was worth it, but am happy to do so

hamishknight replied:
Went ahead and changed this

hamishknight commented:
@swift-ci please test

hamishknight committed:
[ASTGen] NFC: Run swift-format

hamishknight committed:
[ASTGen] Generate type annotations for bindings

Handle the wrapping of TypedPatterns when a type
annotation is specified, and the implicit wrapping
behavior for previously written NamedPatterns.

hamishknight committed:
[ASTGen] Generate AccessorDecls

hamishknight committed:
[AST] Factor out `AccessorDecl::createParsed`

Move the logic for splicing together the
accessor's parameter list out of the parser so
that ASTGen will be able to use it too.

hamishknight committed:
[ASTGen] Better handle ParamDecl arg/param name generation

Move the logic onto the ASTGen side, and plumb
through the flag that will allow subscripts
to have a different argument name defaulting
behavior.

hamishknight committed:
[ASTGen] Better handle pattern binding entries

Generate all entries, and ensure we introduce a
PatternBindingInitializer context for non-local
cases. Also use this opportunity to cleanup
`PatternBindingDecl::create`.

hamishknight committed:
[AST] Introduce `TypedPattern::createPropagated`

hamishknight committed:
[AST] Use `PatternBindingDecl::createImplicit` in more places

And simplify the implementation of `create`.

hamishknight committed:
[ASTGen] Add VarDecls to member contexts

Ensure we add any bound vars to a type or extension
context.

hamishknight committed:
[AST] Compute SubscriptDecl EndLoc

Instead of having the parser set it, compute it
using the recorded brace location for the accessor
record.

hamishknight committed:
[AST] Rename `SubscriptDecl::create` -> `createParsed`

And generate the DeclName using the argument labels
on the ParameterList.

hamishknight committed:
[ASTGen] Generate SubscriptDecls

hamishknight commented:
@swift-ci please test

hamishknight commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -288,23 +288,220 @@ extension ASTGenVisitor {
+      return .willSet
+    case .unsafeAddress:
+      return .address
+    case .addressWithOwner:
+      return .address
+    case .addressWithNativeOwner:
+      return .address
```
Ah, we ought to diagnose these (same for mutable)

hamishknight commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -288,23 +288,220 @@ extension ASTGenVisitor {
+      return .modify
+    case .`init`:
+      return .`init`
+    default:
+      fatalError("Should have diagnosed this")
```
I don't actually think we should be fatal-erroring here

rintaro commented on include/swift/AST/ASTBridging.h:
```diff
@@ -371,25 +378,67 @@ void BridgedDiagnostic_fixItReplace(BridgedDiagnostic cDiag,
+#include "swift/AST/AccessorKinds.def"
+};
+
+struct BridgedAccessorRecord {
```
I was thinking like literally bridging `AbstractStorageDecl::AccessorRecord` and make`AbstractStorageDecl::setAccessors` accept `AbstractStorageDecl::AccessorRecord`.
But considering the current Bridging facility, this might be simpler.

hamishknight replied:
Yeah that was my feeling too

rintaro reviewed (approved):
Nice!

hamishknight commented:
@swift-ci please test

hamishknight committed:
[ASTGen] Be defensive over nil `firstName`

If we only have one name, it should be in
`firstName`, but if it's in `secondName`, handle it
the same, since that's syntactically the same thing.

hamishknight committed:
[test] Split ASTGen tests

hamishknight committed:
[ASTGen] Diagnose unknown accessor specifiers

hamishknight commented:
@swift-ci please test
